src/plugins/kibana_usage_collection/server/plugin.ts

```diff
@@ -72,6 +73,11 @@ export class KibanaUsageCollectionPlugin implements Plugin {
     this.uiSettingsClient = uiSettings.asScopedToClient(savedObjectsClient);
     core.metrics.getOpsMetrics$().subscribe(this.metric$);
     this.coreUsageData = core.coreUsageData;
+    migrateTransactionalDocs(this.logger.get('application-usage'), this.savedObjectsClient).catch(
```

Comment:
I'm a bit concerned that we might fail to migrate them (i.e.: 2+ instances start at similar times and they make each other fail). We won't retry the operation until the next restart.

Reply:
Looking at the implementation of `rollDailyData`, I'm ensure to see how we could encounter failures caused by concurrency (we `bulkCreate` with `overwrite:true`). The real problem in a race here would be that an instance starts the call to `savedObjectsClient.find` while the other one is awaiting for the not-bulk delete of the transactional documents, as it would perform a second, overwriting rolling with only a part of the transactional documents from the first instance. But this was already the case before this change.

Now, I can easily revert this one-time call to a 24h (or any other interval) `timer` instead, wdyt?

Reply:
The errors might come in the calls `savedObjectsClient.delete`. They can get conflicts when trying to delete an object that the other instance already has deleted. Ideally, it will leave only one standing. But that's not ensured in any way.

Reply:
We could migrate the rollup logic to be _safer_ so: 

1. It runs the `find`, then it loops for each document:
2. Tries to delete 1 document (if errored, go back to 1, otherwise continue to step 3)
3. Calls `savedObjectsClient.incrementCounter`.

It'll be much slower, but much safer, and we can ensure there are no duplicates, and that we can run it only at start.

What do you think?

Reply:
> What do you think?

Honestly 😄 ? that any implementation not using a transactional DB and/or an atomic operation will have concurrency issues.

Now, more constructively, the suggested implementation would have the same kind of issues, just the opposite way: what if the `incrementCounter` call fails for any reason, as we already removed the document? 

Also the performance degradation seems significant. We go from 1 `find` x `delete` and 1 `incrementCounter` per appId/dayId tuple to  1 `find` x `delete` and x `incrementCounter`, x being the number of transactional documents.

> The errors might come in the calls savedObjectsClient.delete. They can get conflicts when trying to delete an object that the other instance already has deleted.

We will not get conflicts by trying to delete a non-existing doc, just a 404. We could choose to just ignore 404 coming from the delete operation. 

I would pragmatically do that: ignore 404s from the deletion, then, as a follow-up, implement the `bulkDelete` operation on the SO client/repository to avoid the edge case where an instance perform the `find` call when another one is currently calling the non-transactional `delete` on each document. That way, I think we'll be suppressing all potential race conditions?

Reply:
yeah we would basically need a two phase commit to accurately count + delete these documents without the risk of losing data which is probably not worth the effort considering that this bug was already present for a while and after this PR we won't need the rollup anymore.

Ignoring 404's will solve most of the problems (there are some occurrences of other failure reasons like connection errors). What's the impact of not rolling up until the next restart? If we have a few clusters who only rollup their data e.g. 4 weeks later when the server restarts is that a problem?

(btw bulkDelete/delete_by_query is also not a transaction, so we can still get a race condition between find and bulkDelete. On the spectrum of NoSQL vs SQL databases, ES is a pretty extreme and provides no consistency outside of a single document 🕺  )

Reply:
@afharo Did the change we talk about in https://github.com/elastic/kibana/pull/94923/commits/58f959113a7995d9f50f6a84ce8d3762eab7b674. TIL, we are already ignoring `404`'s in the SOR `delete` implementation, so I just addd the roll until success logic. PTAL.